docs/guide.md

```diff
@@ -0,0 +1,575 @@
+      }
+    }
+  }
+}
+```
+
+# Bolt Cookbook
+
+The rest of this guide will provide sample recipes to solve typical problems that developers
+face in securing their Firebase databases.
+
+## Dealing with Timestamps
+
+You can write timestamps (Unix time in milliseconds) in Firebase data and ensure that whenever
+a time is written, it exactly matches the (trusted) server time (independent of the clock on
+the client device).
+
+```javascript
+path /posts/$id is Post;
+
+type Post {
+  // Make sure that the only value allowed to be written is now.
+  validate() = this.modified == now;
+
+  message: String,
+  modified: Number
+}
+```
+
+Each time the Post is written, modified must be set to the current time (using
+[ServerValue.TIMESTAMP](https://www.firebase.com/docs/web/api/servervalue/timestamp.html)).
+
+A handy way to express this is to use a user-defined type for the Modified timestamp:
+
+```javascript
+path /posts/$id is Post {
+  read() = true;
+  write() = true;
+}
+
+type Post {
+  message: String,
+  modified: Modified
+}
+
+type Modified extends Number {
+  validate() = this == now;
+}
+```
+
+Similarly, if you want to have a Created property, it should match the current time
+when first written, and then never changed thereafter:
+
+```javascript
+path /posts/$id is Post {
+  read() = true;
+  write() = true;
+}
+
+type Post {
+  message: String,
+  modified: Modified,
+  created: Created
+}
+
+type Modified extends Number {
+  validate() = this == now;
+}
+
+type Created extends Number {
+  validate() = initial(this, now);
+}
+
+// Returns true if the value is intialized to init, or retains it's prior
+// value, otherwise.
+initial(value, init) = value == (isInitial(value) ? init : prior(value));
+isInitial(value) = prior(value) == null;
+```
+
+Note the special function `prior(ref)` - returns the previous value stored at a given database location
+(only to be used in validate() and write() rules).
+
+```JSON
+{
+  "rules": {
+    "posts": {
+      "$id": {
+        ".validate": "newData.hasChildren(['message', 'modified', 'created'])",
+        "message": {
+          ".validate": "newData.isString()"
+        },
+        "modified": {
+          ".validate": "newData.isNumber() && newData.val() == now"
+        },
+        "created": {
+          ".validate": "newData.isNumber() && newData.val() == (data.val() == null ? now : data.val())"
+        },
+        "$other": {
+          ".validate": "false"
+        },
+        ".read": "true",
+        ".write": "true"
+      }
+    }
+  }
+}
+```
+
+## Timestamped Generic (parameterized) Types
+
+Bolt allows types to be parameterized - much like Java Generic types are defined.  An alternate way
+to define the Timestamp example above is:
+
+```javascript
+// Note the use of Timestamped version of a Post type.
+path /posts/$id is Timestamped<Post> {
+  read() = true;
+  write() = true;
+}
+
+type Post {
+  message: String,
+}
+
+type Timestamped<T> extends T {
```

Comment:
T could be used anywhere in the type definition...in this example, it's used in the _extends_ base type - so a Timestamped<T> is a "subclass" of a T.  You could also do
```
type Timestamped<T> {
  data: T,
  modified: Modified
}
```